File: problems/insert_delete_getrandom_o(1)/solution.cpp
//see python implememntation also, that is easy


// This problem requires us to design a data structure that supports insert, remove, and getRandom in constant time O(1).

// This solution uses two data structures, a Hash Table and a resizable array. The Hash Table maps the values to their respective indices in the array.

// Insert Operation
// This operation can be easily supported by inserting the new value to the end of the resizable array and updating its index in the Hash Table.

// Remove Operation
// This operation is a little tricky. We can easily remove any element from the Hash Table in constant time as long as we know it's value. However, removing an element from the middle of the array is a costly operation. Fortunately, removing the last element of the array can be done in constant time ( using pop_back() ). We will swap the value we want to remove with the last element of the array and then remove the last element of the array.

// getRandom Operation
// This operation can be easily done by using C++ inbuilt rand() on the array.

// The code is as follows.

class RandomizedSet {
    vector<int> numberList;
    unordered_map<int, int> hashTable;
public:
    /** Initialize your data structure here. */
    RandomizedSet() {}
    
    /** Inserts a value to the set. Returns true if the set did not already contain the specified element. */
    bool insert(int val) {
        if(hashTable.find(val) != hashTable.end())
        {
            return false;
        }
        hashTable[val] = numberList.size();
        numberList.push_back(val);
        return true;
    }
    
    /** Removes a value from the set. Returns true if the set contained the specified element. */
    bool remove(int val) {
        if(hashTable.find(val) == hashTable.end())
        {
            return false;
        }
        swap(numberList[hashTable[val]], numberList[numberList.size() - 1]);
        
        numberList.pop_back();
        
        //now update key value pairs of both last and val in hashmap
        
        //updating index of last(that is currently at index of val) to new index ( that is old index of val in numberList, before swapping, we have this info in hasmap as hashTable[val])
        hashTable[numberList[hashTable[val]]] = hashTable[val];   //numberList[hashTable[val]] gives last element, as eleemnt at old ind of val before swapping, we have this info in hasmap as hashTable[val])
        //removing val
        hashTable.erase(val);
        return true;
    }
    
    /** Get a random element from the set. */
    int getRandom() {
        return numberList[rand() % numberList.size()];
    }
};

File: problems/insert_delete_getrandom_o(1)/solution.py
# We just keep track of the index of the added elements, so when we remove them, we copy the last one into it.

# From Python docs (https://wiki.python.org/moin/TimeComplexity) we know that list.append() takes O(1), both average and amortized. Dictionary get and set functions take O(1) average, so we are OK.


class RandomizedSet(object):

    def __init__(self):
        self.nums, self.pos = [], {}
        
    def insert(self, val):
        if val not in self.pos:
            self.nums.append(val)
            self.pos[val] = len(self.nums) - 1      # 0 based indexing
            return True
        return False
        
#most jod thing i have ever seen !

    def remove(self, val):
        if val in self.pos:
            idx, last = self.pos[val], self.nums[-1]
            self.nums[idx] = last          # update val to last element (and we will pop the last now bcz we replaced val by last, now we dont have val in nums. So we also need to remove val from map also and update index of last in map also)
            self.pos[last] = idx           # updating index of last in map, new index of last will be old index of val that is pos[val] (idx)
            
            self.nums.pop()                                  # remove last element from nums
            self.pos.pop(val)                                # remove val from map
            return True
        
        return False
            
        
    def getRandom(self):    # random.randInt chooses a random value frokm given range of indices, and we then return value at that index
        return self.nums[random.randint(0, len(self.nums) - 1)]